Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp

```diff
@@ -0,0 +1,268 @@
+		convertCSVListToVector(optimizeFormulaNamesPL->get(PT), scoreNames);
+	}
+
+
+	epsilon = epsilonPL->get(PT);
+	epsilonByRange = epsilonByRangePL->get(PT);
+	tournamentSize = tournamentSizePL->get(PT);
+	nextPopSizeFormula = stringToMTree(nextPopSizePL->get(PT));
+	numberParents = numberParentsPL->get(PT);
+	recordOptimizeValues = recordOptimizeValuesPL->get(PT);
+	
+	// leave this undefined so that max.csv is not generated
+	//optimizeFormula = optimizeValueMT;
+
+	popFileColumns.clear();
+	if (recordOptimizeValues){
+		for (auto name : scoreNames) {
+			popFileColumns.push_back(name);
+		}
+	}
+}
+
+
+int LexicaseOptimizer::lexiSelect(const std::vector<int> &tournamentIndexList) {
+	if (!scoresHaveDelta) { // if all scores are the same! pick random
+		return Random::getIndex(tournamentIndexList.size());
+	}
+
+	// generate a vector with formulasSize and fill with values from 0 to formulasSize in a random order.
+	std::vector<int> formulasOrder;
+	std::vector<int> sourceValues; // temp vector used to generate formulasOrder
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		sourceValues.push_back(i);
+	}
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		int pick = Random::getIndex(sourceValues.size());
+		formulasOrder.push_back(sourceValues[pick]);
+		sourceValues[pick] = sourceValues.back();
+		sourceValues.pop_back();
+	}
+	// now we have a random order in formulasOrder
+
+	// keepers is the current list of indexes into population for orgs which have passed all tests.
+	std::vector<int> keepers  = tournamentIndexList;
+
+	while (keepers.size() > 1 && formulasOrder.size() > 0) {
+		// while there are still atleast one keeper and there are still formulas
+		int formulaIndex = formulasOrder.back();
+		formulasOrder.pop_back();
+
+		double scoreCutoff;
+
+		if (epsilonByRange) {
+			// get the range of scores
+			double maxScoreHere = scores[keepers[0]][formulaIndex];
+			double minScoreHere = maxScoreHere;
+			for (size_t i = 0; i < keepers.size(); i++) {
+				maxScoreHere = std::max(scores[keepers[i]][formulaIndex], maxScoreHere);
+				minScoreHere = std::min(scores[keepers[i]][formulaIndex], minScoreHere);
+			}
+			// get scoreCutoff relivite to min and max
+			scoreCutoff = (maxScoreHere - ((maxScoreHere - minScoreHere) * epsilon));
+		}
+		else {
+			// create a vector of remaning scores and sort them
+			std::vector<double> keeperScores;
+			for (size_t i = 0; i < keepers.size(); i++) {
+				keeperScores.push_back(scores[keepers[i]][formulaIndex]);
+			}
+			std::sort(keeperScores.begin(), keeperScores.end());
+			// scoreCutoff is the (size of vector * epsilon) + 1 from the back
+			scoreCutoff = keeperScores[keeperScores.size()
+				- static_cast<int>(((static_cast<double>(keeperScores.size())) * epsilon) + 1)];
+		}
+		// for each keeper, see if there are still a keeper, i.e. they have score >= scoreCutoff
+		for (size_t i = 0; i < keepers.size();) {
+			if (scores[keepers[i]][formulaIndex] >= scoreCutoff) {
```

Comment:
This should be a std::remove_if

Reply:
this would require a redesign of the scores data structure since I need to remove from the outer vector based on the inner vector. a later refactoring should be considered.